lib/krb5/pac.c

```diff
@@ -1200,7 +1202,17 @@ _krb5_pac_sign(krb5_context context,
 		ret = krb5_enomem(context);
 		goto out;
 	    }
-	    /* XXX if not aligned, fill_zeros */
+
+	    if (p->pac->buffers[i].type == PAC_LOGON_INFO
```

Comment:
Is there a reason the alignment is only for `PAC_LOGON_INFO` and `PAC_UPN_DNS_INFO`? If so, could you put a reference to the protocol docs in a comment?

Reply:
I based this on this bug report, which mentions those buffer types explicitly: [https://bugzilla.samba.org/show_bug.cgi?id=13134](https://bugzilla.samba.org/show_bug.cgi?id=13134)

However, it may be the case that Windows aligns the other PAC buffers as well.

Reply:
Perhaps check with dochelp and, if it turns out they are all aligned, then make this a general case?

Reply:
As far as I can see this is about padding not alignment.

MS-PAC 2.4 PAC_INFO_BUFFER says:
...
cbBufferSize (4 bytes): A 32-bit unsigned integer in little-endian format that contains the size, in
bytes, of the buffer in the PAC located at Offset.
...
Offset (8 bytes): A 64-bit unsigned integer in little-endian format that contains the offset to the beginning of the buffer, in bytes, from the beginning of the PACTYPE structure (section 2.3). The data offset MUST be a multiple of eight. The following sections specify the format of each type of element.
...


So Offset contains the alignment and that was already aligned to 8 bytes, this change is about
cbBufferSize and I think asking dochelp would be good, with having additional tests (with the python krb5
infrastructure in Samba), we should test with different sAMAccountName length values or account relevant fields
on the account object.

I'm also not sure if this is really the correct place for that change?

Shouldn't p->pac->buffers[i].buffersize already have the padded length?
Which would mean the Samba part should add the padding...

Reply:
Looking at it I'm wondering if the "/* advance data endpointer and align */" section below is really correct,
I'd guess the alignment part of it should be moved to the top of the loop body, so that we don't add unused padding
at the end of the last loop iteration.

Reply:
Avoiding unneeded padding is good (given ticket size limits). First thing to fix though would be removing special cases for padding.

Reply:
I ran tests against Windows and this change seems to match what Windows is doing. LOGON_INFO and UPN_DNS_INFO are padded to a multiple of 8 bytes, while CREDENTIAL_INFO, the checksum types, LOGON_NAME, and REQUESTER_SID are not. The CONSTRAINED_DELEGATION and ATTRIBUTES_INFO blobs seem to already be multiples of 8 bytes.

Reply:
Ah ok, thanks for checking!

Reply:
> I ran tests against Windows and this change seems to match what Windows is doing. LOGON_INFO and 
> UPN_DNS_INFO are padded to a multiple of 8 bytes, while CREDENTIAL_INFO, the checksum types, LOGON_NAME, 
> and REQUESTER_SID are not. The CONSTRAINED_DELEGATION and ATTRIBUTES_INFO blobs seem to already 
> be multiples of 8 bytes.

Why didn't you change the Samba code to add the padding? I think that would be the correct place, which means we'd also have it for the MIT KDC case...

Reply:
Yeah, I think this is the right spot actually.  Sorry for the stuffing about @lhoward